Adding options cssClassTabsLi and cssClassTabsA

robsdedude opened the pull request:
See my comment [here](https://github.com/mikejacobson/jquery-bootstrap-scrolling-tabs/issues/17#issuecomment-537027214).
These options make using this nice widget much easier with BootStrap4. I've also included a new example in run to demonstrate and test this.

One thing: Since bower does not support multiple versions of the same dependency (at least I couldn't find out how), I'm afraid the example in run will not work when the libs are installed using bower. Generally I couldn't quite find out how bower is supposed to act in this repo.

robsdedude committed:
Adding options cssClassTabsLi and cssClassTabsA

robsdedude commented:
If this pull request gets merged I'd like to submit another one that outlines the steps to use this lib with BS4 in the README.

mikejacobson commented on dist/jquery.scrolling-tabs.js:
```diff
@@ -1246,20 +1246,32 @@
+                .attr('href', '#' + tab[propNames.paneId])
+                .html(tab[propNames.title]);
+      if (options['cssClassTabsA']) {
+          $a.addClass(options['cssClassTabsA']);
```
Instead of introducing new option `cssClassTabsA`, what do you think about just checking the existing BS4 flag, and if it's `true`, adding the `nav-link` class directly since we already know that's the class required by BS4?

Your approach makes it more flexible but requires the user to provide two more settings which I expect would always be the same for BS4, so it would simplify the API if we just hardcode those values.

What do you think?

robsdedude replied:
By the time opening the PR, I wasn't aware of the bootsrap4 option. So yes, I will.

mikejacobson commented on package.json:
```diff
@@ -20,6 +20,7 @@
   ],
   "dependencies": {
     "bootstrap": "^3.1.1",
+    "bootstrap4": "npm:bootstrap@^4.3.1",
```
Can you remove this. I wouldn't want to introduce BS4 as a dependency for current users who are only using BS3.

robsdedude replied:
I can. But it'll break the demo in `run`. So I'd have to remove it as well.

robsdedude replied:
I moved it to the dev dependencies. How do you like that approach?

robsdedude committed:
Infer BS4 css classes from bootstrapVersion option

robsdedude committed:
Fixed further bs4 incompatibilities

also adjusted dependencies:
bootstrap 3 || 4 for the lib
and bs3 AND bs4 as dev dependencies for the demos

mikejacobson commented on src/js/tabListeners.js:
```diff
@@ -274,23 +274,45 @@ function listenForDropdownMenuTabs($scroller, stc) {
+        $ddParentTabLi
+          .add($selectedMenuItemLi)
+          .find('a')
+          .addClass('active');
```
Well-done here in this entire `if` block! 👍 I had to walk myself through it a couple times to make sure I understood what all was going on. Looks really good, though. I trust that it actually works correctly. 😄

robsdedude replied:
I just added another example in `run` to check that it works.

BTW it might be worth considering to rename `run` -> `examples` so the devs using the package have an easier time finding the examples.

mikejacobson reviewed (changes requested):
I had a few comments. Thanks for all the work! Overall it looks really good. You obviously put a lot of effort into this.

robsdedude commented on src/js/_main.js:
```diff
@@ -1,8 +1,19 @@
+    typeof jQuery.fn.tab.Constructor === 'undefined' ||
+    typeof jQuery.fn.tab.Constructor.VERSION === 'undefined') {
+  throw new Error("jquery-bootstrap-scrolling-tabs requires Bootstrap's tab plugin for jQuery");
+}
```
The problem with 3 as default without a means to check if 4 is installed is that `"dependencies": {
    "bootstrap": "^3.1.1 || ^4.3.1", …}` will cause the package manager to install bs4 unless the developer installed bs3 before. So I think it's problematic to choose 3 as default while the package manager by default will install 4.

robsdedude replied:
How about I remove the check if BS exists but try to find `jQuery.fn.tab.Constructor.VERSION` and if it exists I use this version as default. If it doesn't exist I'll use 3.

robsdedude commented on src/js/api.js:
```diff
@@ -1,3 +1,12 @@
+  jQuery.fn &&
+  jQuery.fn.tab &&
+  jQuery.fn.tab.Constructor &&
+  jQuery.fn.tab.Constructor.VERSION &&
```
I'll resolve this as we continue the discussion [here](https://github.com/mikejacobson/jquery-bootstrap-scrolling-tabs/pull/73#discussion_r334743955).

robsdedude committed:
Fix typo in func name

robsdedude commented on src/js/buildTabs.js:
```diff
@@ -43,26 +43,42 @@ var tabElements = (function () {
     var liContent = options.tabLiContent || '<li role="presentation" class=""></li>',
         $li = $(liContent),
-        $a = getNewElTabAnchor(tab, propNames).appendTo($li);
+        $a = getNewElTabAnchor(tab, propNames, aOptions).appendTo($li);
```
I stuck to the code style of yours. At many other locations you extracted only the needed options for subroutines to pass them on. So I did the same here. If it were my code base I would have passed the whole options object through all the functions. It saves you future work if you decide to extend the options.

robsdedude commented on src/js/tabListeners.js:
```diff
@@ -274,23 +274,45 @@ function listenForDropdownMenuTabs($scroller, stc) {
+      if (options.bootstrapVersion == 4) {
+        isActive = $selectedMenuItemLi.hasClass('active');
+      } else {
+        isActive = $selectedMenuItemLi.find('a').hasClass('active');
```
Good catch. The check is indeed inverted.

I consider multi-line ternary operators an anti-pattern. The save you 2 lines of space but make the code more confusing. But if you insist I'll change it.

robsdedude committed:
[BS4] Fixed small bugs and code style things

robsdedude commented on src/js/api.js:
```diff
@@ -152,10 +166,12 @@ $.fn.scrollingTabs.defaults = {
   leftArrowContent: '',
   rightArrowContent: '',
+  cssClassTabsLi: null,
+  cssClassTabsA: null,
```
I could. But do more options hurt?

robsdedude commented on src/js/bsVersionUtil.js:
```diff
@@ -0,0 +1,44 @@
+    return $scroller.find('li[role="presentation"] > a[role="tab"].active');
+  } else {
+    return $scroller.find('li[role="presentation"].active');
+  }
```
I don't like multi-line ternary operators. I think it makes code harder to read.

robsdedude committed:
Using more ternary operators

robsdedude commented on src/js/api.js:
```diff
@@ -6,6 +15,11 @@ var methods = {
+        console.warn('bootstrapVersion =', options.bootstrapVersion,
+                     'is unsupported. Falling back to 3.');
+        options.bootstrapVersion = 3;
+    }
```
I'll remove it. But let me comment on this: I hate the typical javascript mentality of silently failing. It causes head scratching developers that can't for heavens sake find out why there code is just not doing things right, while the fix could be as easy as fixing a little typo but the lib just doesn't tell the dev that it received some unexpected input.

robsdedude committed:
On request: remove bs version check in options

robsdedude commented:
Ok. Please check my comments and latest changes. I'll leave the PR as is. I'd be happy if you were able to take it from here (shouldn't be much work left to be done).
